### PR TITLE
Feature/caching

### PR DESCRIPTION
**Pull request by iliiliiliili**
Add caching mechanism

Call `dataset.cached()` to be cached in `.datasetops_cache` folder.
Call `dataset.cached(path)` to be cached in `path` folder.

Cache folder contains `database.json` and `cache_*.pkl` files.
To clear cache folder use `Cache.clear()` or `Cache.clear(path)`.

`database.json` file contains list of entries, where each entry has an `identifier` and `cache_id`.
`cache_{num}.pkl` stores data for entry with `cache_id=num`. It is generated using `dill`.
`identifier` is a `base64` string of `dill.dumps` data built on `TransformationGraph` for the dataset.
`TransformationGraph` is built by making a graph based on `_get_origin` values of each dataset. This graph as acyclic but not always a tree.
Each operation on the dataset describes result dataset as created using this operation.
Some operations that require `seed` can set `dataset.cachable` to `False` if `seed` is `None`.
Loaders that load from data from memory and don't provide `identifier` also set `dataset.cachable` to `False`.

### CODE REVIEW

File: tests/datasetops/test_caching.py
Comment (LukasHedegaard) [inline tests/datasetops/test_caching.py]: We could consider adding automatic readers for txt and bin data as part of the library

**Review by LukasHedegaard — CHANGES_REQUESTED** [review]
Hey, good work on this, I think the caching mechanism itself is sound.

A few caveats, though: 
I think the transformation graph may be unnecessarily complicated
- Why do we need to hash the whole object? I think we should instead use the information on how it was created. Otherwise, it might get unnecessarily slow.

Running the tests has gotten incredibly slow.

For comparison running the `test_datasets` (which contains most of the test until now) runs in 5 seconds (this includes loading time for tensorflow and training a small neural network).

The `test_transformation_graph` takes the same time. I believe this is due to the deep copies and hashing of dataset objects (the object internals are hashed too, no?). 
- Also, it seems like a there are a lot of redundant test - could we discard of the redundant ones?

The main test-time issue is with the `test_caching`, which takes a full minute!
- I think the issue is mainly that many of the tests are repeated. For instance, doesn't the second test below only test one additional value as compared to the first? It performs the caching for all four datasets, though. As far as I can see, assert_cache doesn't compare the datasets with one another, so what is the point?

        test2 = test1.image(False, True, False)
        assert_cache([test2, test1, test], cache_path)

        test3 = test2.transform((read_text, None, None))
        assert_cache([test3, test2, test1, test], cache_path)

Before merging, I think we should 
1) Remove redundant tests (we should aim for less than 5 sec per test)
2) Maybe reconsider the transformation graph. I think we can achieve the same in a more performant, simpler and human-readable manner
3) Add some benchmarking to ensure that the caching actually yields a performance boost (otherwise we shouldn't include it)

**Comment by iliiliiliili** [comment]
Hello Lukas,
Thanks for the review.
1. Those tests are needed to be sure that in one "session" different datasets don't intersect and their identifiers are truly unique.
2. `TransformationGraph` indeed stores links for `Dataset` objects, but serialized values don't include them, `serialize` method calls internal function `minimize` that filters unnecessary for serialization data from the graph.
I have made serializing mechanism in the way, that identifier is not a hash, but an encoded version of transformation graph, so, having cache identifier and original data, it is possible to restore the code of the dataset without the original code. If this becomes a problem, we can just hash this identifier, but in current implementation there is zero possibility for 2 different datasets to have same identifiers.
3. I will add benchmarking in separate branch.

File: tests/datasetops/test_caching.py
Comment (iliiliiliili) [reply tests/datasetops/test_caching.py]: Yes, we can do this. While doing it, it would be good to add tests, so they will use both internal and user-defined readers

**Comment by LukasHedegaard** [comment]
> 1. Those tests are needed to be sure that in one "session" different datasets don't intersect and their identifiers are truly unique.

In each `assert_cache`, all the datasets are cached, tested against the version prior to cache, read from cache anew and tested against the original again. Do you need to repeat this every time a new one is added? There are 18 datasets, which you test, both with default cache folder and a custom one. So, there, there are 2*(18+17+16+...+2+1) = 342 runs of [cache, test, load cache, test]. 
I would think we can the the same basic logic with 18 + 1 [cache, test, load cache, test] and a single step for testing that there a no clashes. This would bring down the test time to below 4 seconds.

> 2. `TransformationGraph` indeed stores links for `Dataset` objects, but serialized values don't include them, `serialize` method calls internal function `minimize` that filters unnecessary for serialization data from the graph.
>    I have made serializing mechanism in the way, that identifier is not a hash, but an encoded version of transformation graph, so, having cache identifier and original data, it is possible to restore the code of the dataset without the original code. If this becomes a problem, we can just hash this identifier, but in current implementation there is zero possibility for 2 different datasets to have same identifiers.

It's definitely nice, that there can be no clashes! I noticed, that you added a `cacheable` flag, which is set to `False` in the regular Loader. Why is this? And can datasets downstream from a non-cacheable loader be cached? Can we distinguish between two different (non-cacheable) Loaders, if the same operations applied to them?
Could the the TransformationGraph also be used to create a nice human readable format?

> 3. I will add benchmarking in separate branch.

Like I said, I think we should make sure that performance improves by using the cache before we merge this branch into develop. Otherwise, we will need to revert the merge later if it turns out that the caching didn't improve performance. If you want to do the benchmarking in a separate branch, we should not merge this branch into develop, but the benchmark branch if the results are good.

**Comment by iliiliiliili** [comment]
> In each assert_cache, all the datasets are cached, tested against the version prior to cache, read from cache anew and tested against the original again. Do you need to repeat this every time a new one is added? There are 18 datasets, which you test, both with default cache folder and a custom one. So, there, there are 2*(18+17+16+...+2+1) = 342 runs of [cache, test, load cache, test].
I would think we can the the same basic logic with 18 + 1 [cache, test, load cache, test] and a single step for testing that there a no clashes. This would bring down the test time to below 4 seconds.

Okay, I will do so.

> I noticed, that you added a cacheable flag, which is set to False in the regular Loader. Why is this? And can datasets downstream from a non-cacheable loader be cached? Can we distinguish between two different (non-cacheable) Loaders, if the same operations applied to them?
Regular Loader has it set to `False` only if `identifier` is not set:
```
if self.identifier is None:
    self.cachable = False
```
For folder-based loaders it's set to the path of the folder (so if folder data is changed, cache should be cleared by user). For memory-based loaders (from pytorch/tf/numpy) user has to specify identifier for it to be cachable. If any of the datasets in transformation graph are uncachable, we are unable to cache this dataset.

One more idea: we can add function to check cache. Based on the identifier we can restore transformation graph and apply it to the original data again and compare with cache, if there is a difference, we can rewrite the cache. It will solve the problem with original data modifications.

>Like I said, I think we should make sure that performance improves by using the cache before we merge this branch into develop. Otherwise, we will need to revert the merge later if it turns out that the caching didn't improve performance. If you want to do the benchmarking in a separate branch, we should not merge this branch into develop, but the benchmark branch if the results are good.

Ok, I think we can start benchmarking here, and if we will need more benchmarking, maybe, unrelated to caching, we can create new branch.

**Comment by LukasHedegaard** [comment]
Okay, sounds good! 👌
A single comment regarding the cacheable flag: in the cache function, we should include a nice Error message stating how the user can make a non-cacheable dataset cacheable (in case it's not there already)

**Comment by iliiliiliili** [comment]
I changed behavior of loading data from cache. Preprocessed `KITTI` train dataset takes 12.5 GB of memory, so storing it in memory not always can be done. I created `StreamDataset` that can use any stream as a source of data (hence we can use `tensorflow` and `pytorch` dataset objects also). Because it is a stream, by default only step-by-step read is allowed, but user can specify `keep_loaded_items=True` to force it to save read items in memory, this mechanism also allows random access to `StreamDataset` (it will read elements one by one until required is read, next time it will read it (and all previous) directly from memory).
Because items can have different size, we can't just `seek` through file stream to get desired item by index, but theoretically, we can try to do it for fixed-size dataset or, for all other datasets by storing additional information about item sizes.

I added benchmarking using preprocessing of the `KITTI` dataset. I have tested it using different amount of elements. `Repeats` means how much data reading cycles are done. `Storage` indicates is dataset is fully saved in memory or not. `Time Raw` means how many seconds it took to traverse original dataset. `Time Cached` is for cached version of same dataset. `Caching time` indicates time for saving dataset to cache file. `Difference` means how much `Time Raw` is greater than `Time Cached`.

image

**Comment by LukasHedegaard** [comment]
Hey, great work on this! Transformation_graph.display is a nice touch :-)
I saw an opportunity to streamline the code a bit, so I've made some changes here and there. Mostly, they were minor modification, moving stuff a bit (e.g. the KITTI dataset location), but I've also added a few additional features (see commit messages).
Moreover, I've made a Black-compatible flake8 setting and updated the GitHub workflow to actually check linting (as seen in Makefile and .github/workflows/pythonpackage.yml).
I could not make the benchmarking work, though. Is this because you used a local version of whole KITTI dataset for this?
Please give the changes a look and check if you agree with them.

**Comment by iliiliiliili** [comment]
Hi Lukas, nice changes, but I disagree at one point:

You optimized origin calculation in `Dataset.__init__` and, as part of it, removed raising an `Exception` when `operation_name` is unknown. I strongly recommend keeping it because all of us are likely to add new operation and can forget to include origin calculations in it. While it's hard to forget to include operation name in constructor, it's easy to forget to put it in right category (like seed-dependent or not, or even create new category). Even for me it helped to debug.
In new code when someone adds an operation that depends on seed, but didn't specify it, we won't see a problem and even will not see it for a long time if they won't add a test with this operation to `test_cachable`.

**Comment by LukasHedegaard** [comment]
Good point, let’s roll it back to raise an error for unknown operations. Would you like to do the honors, or should I? And btw, did you see my comment regarding benchmarks not running?

**Comment by iliiliiliili** [comment]
You can do it :)
About benchmarks: yes, it requires full KITTI dataset, so I didn't want to increase the size of the project by 30 GB :)
I think, we can make it more beautiful later, and, maybe, add downloader for KITTI with warning about its size.

**Review by LukasHedegaard — APPROVED** [review]
I'm happy with the branch now. The caching is definitely nice addition. Feel free to merge it into master, if you are satisfied as well @iliiliiliili

**Pull request merged**
